Integrate kpi-calculator from PyPI

jarsarasty opened the pull request:
- Adds `kpi-calculator>=0.4.1` as a dependency
- Uses `calculate_kpis_from_simulator` and `build_esdl_string_with_kpis` to compute KPIs and embed the results into the output ESDL string
- Supports optional workflow config parameters: `system_lifetime` (years), `discount_rate` (%), and `round_up_replacement` (bool, default `True` for MESIDO optimizer compatibility)
- KPI failure is non-fatal: any exception is logged and the simulation result is returned without KPIs
- Adds an integration test

jarsarasty commented on unit_test/test_kpi_integration.py:
```diff
@@ -0,0 +1,201 @@
+                ("port_1", "pressure"): [100000.0, 100100.0, 100200.0],
+                ("port_2", "temperature"): [290.0, 291.0, 292.0],
+            },
+            index=pd.date_range("2024-01-01", periods=3, freq="h"),
```
For reproducibility of tests, we will keep the hardcoded dates.

jarsarasty commented on unit_test/test_kpi_integration.py:
```diff
@@ -0,0 +1,201 @@
+            input_esdl = f.read()
+
+        # Configure workflow - use Unix timestamps (seconds since epoch)
+        import datetime
```
Minor style issue

jarsarasty commented on pyproject.toml:
```diff
@@ -72,9 +73,11 @@ starting_version = "0.0.1"
 
 [tool.coverage.run]
 source = ["src"]
+omit = ["unit_test/test_kpi_integration.py"]
```
Irrelevant: we don't measure test coverage of test files.

jarsarasty commented:
This PR is now updated with the new pre-release of the KPI calculator.

jarsarasty committed:
feat: upgrade kpi-calculator to 0.4.1 and integrate KPI output into simulator worker
- Bump kpi-calculator to 0.4.1
- Pass simulation results to calculate_kpis_from_simulator
- Embed KPI results in output ESDL via build_esdl_string_with_kpis
- Support optional system_lifetime, discount_rate, and round_up_replacement config params
- Add end-to-end integration test verifying CAPEX, OPEX, and energy KPIs

samvanderzwan commented on unit_test/test_kpi_integration.py:
```diff
@@ -0,0 +1,89 @@
+            )
+
+        # Verify output ESDL structure
+        assert output_esdl is not None
```
In the simulator we are not using assert, but self.assertAlmostEqual and similar for the checks

jarsarasty replied:
All bare assert statements in the tests have been replaced with self.assertIn, self.assertAlmostEqual, and other unittest assertion methods

samvanderzwan commented on unit_test/test_kpi_integration.py:
```diff
@@ -0,0 +1,89 @@
+        # Verify output ESDL structure
+        assert output_esdl is not None
+        esh = pyesdl_from_string(output_esdl)
+        energy_system = esh.energy_system
```
It seems you are doing acts and asserts step by step. In the simulator we try to follow the triple A pattern:
Arrange, Act, Assert. In the Act we only do one thing, the thing the test is for. In this test it seems like you are testing multiple things. This might give you problems later when it fails.

jarsarasty replied:
Each test now follows the Arrange/Act/Assert pattern with a single assertion per test method. The setUpClass runs the simulation once and stores results in cls.kpi_by_name, and each test checks exactly one thing against that result

samvanderzwan commented on src/simulator_worker/simulator_worker.py:
```diff
@@ -111,8 +128,31 @@ def simulator_worker_task(
+        kpi_results = calculate_kpis_from_simulator(
+            result_indexed,
+            input_esdl,
+            **({"system_lifetime": system_lifetime} if system_lifetime is not None else {}),
```
What happens when it is None?

jarsarasty replied:
output_esdl is set by create_output_esdl. It always holds the simulation ESDL with results. The KPI block produces a separate output_esdl_with_kpis: if build_esdl_string_with_kpis succeeds it holds the ESDL with kpis, and if it raises the except block explicitly falls back to output_esdl_with_kpis = output_esdl (with simulation results, no KPIs).

samvanderzwan commented on src/simulator_worker/simulator_worker.py:
```diff
@@ -57,9 +71,12 @@ def simulator_worker_task(
+    - start_time: float, seconds since epoch
+    - end_time: float, seconds since epoch
+    - timestep: float, seconds
+    - system_lifetime: float (optional), system lifetime in years for KPI calculation
```
We need to discuss the use of the unit here. In the simulator we said everything in SI. But here it seems logical to have years.

jarsarasty replied:
Good point for discussion. The SI unit for time is seconds, but system_lifetime as a concept is always expressed in years in both the kpi-calculator API (`DEFAULT_SYSTEM_LIFETIME_YEARS`) and the user-facing config. Passing seconds would be confusing and error-prone. We propose keeping years here as a deliberate exception, documented in the docstring. Open to alternative approaches if the team prefers strict SI.

samvanderzwan commented on src/simulator_worker/simulator_worker.py:
```diff
@@ -43,6 +45,18 @@
 logger = logging.getLogger("simulator_worker")
 
 
+def _parse_bool_config(config: ProtobufDict, key: str, default: bool) -> bool:
```
please add tests for these functions

jarsarasty replied:
Done — _parse_bool_config and _parse_float_config are now covered by unit tests in test_hello.py (the existing utils test file), and save_debug_esdl is exercised via the integration test with debug_esdl=True.

samvanderzwan reviewed (changes requested):
Some small comments

jarsarasty committed:
Improve KPI integration: config parsing, debug artifacts, and test quality

  - Bump kpi-calculator to >=0.4.2a1 to get DEFAULT_SYSTEM_LIFETIME_YEARS
    and DEFAULT_DISCOUNT_RATE_PERCENT from the public API
  - Add save_debug_esdl() to capture input/output ESDL files per run;
    opt-in via debug_esdl workflow config key
  - Upload debug ESDL as CI artifact on every test run for inspection
  - Move _parse_bool_config/_parse_float_config to utils.py with proper
    defaults and warnings when system_lifetime or discount_rate are absent
  - Rewrite integration tests with unittest.TestCase, AAA pattern, and
    setUpClass; add TestAllKPICategories covering all KPI categories
    (costs, NPV, energy, CO2 emissions) with debug artifact enabled

jarsarasty committed:
Update kpi-calculator to v0.4.2a2

Bump kpi-calculator from 0.4.2a1 to 0.4.2a2 in pyproject.toml,
requirements.txt, and dev-requirements.txt. Save ESDL files
from integration tests to a temp dir with automatic cleanup.
Suppress the parameterized mypy warning via config instead
of a per-line type: ignore comment.

jarsarasty commented:
> Some small comments

Thank you for your comments, @samvanderzwan. 
The PR is now updated

samvanderzwan reviewed (approved):
lgtm
